java-sdk/sdk-checkpoints.asciidoc

```diff
@@ -0,0 +1,88 @@
+
+A complete item is an item that was emitted first as a candidate and sent to the fetcher to be processed. The fetcher completes the process by sending a FetchResult message to indicate the process is finished. This item is then marked as complete by the SDK controllers by setting the `blockID` field in the item metadata to match the blockID of the current job.
+
+The blockId is used to quickly identify items in the CrawlDB which may not have been fully processed (completed). A blockId identifies a series of one or more Jobs. The lifetime of a blockId spans from the start of a crawl to the completion. The SDK controller will generate and use a new blockId when:
+
+* The current job is the first job.
+* The previous job’s state is `FINISHED`.
+
+When a job starts and the previous job did not complete, the job's state is `STOPPED`. In this case, the previous job's blockId is reused. The same blockId will be reused until the crawl successfully completes. The SDK controller will continue checking the CrawlDB for incomplete items, which are identified by having a blockID that doesn't match the previous job blockID. This approach ensures all items within the job are completed before the next job beings, even if the job was stopped multiple times before completion.
+
+When an item is considered a candidate, the item's blockId is not added to, or updated in, the item’s metadata. When the item is processed by the fetcher, and a FetchResult was emitted for that item, the blockId is added to the item metadata and stored in the SDK CrawlDB. The item is then marked as complete and will be processed in the next non-stopped job.
+
+When all items are complete, the SDK will check for checkpoints, as detailed in <<checkpoint-design, Checkpoint Design>>.
+
+IMPORTANT: A job won’t be started if there are incomplete items from the previous job stored in the SDK CrawlDB.
+
+===== Item Completion Flow
+image:/assets/images/sdkcheck-itemcomplete.png[Item Completion Flow]
+
+. The SDK controller gets a `FetchInput` from the SDK CrawlDB and sends it to the fetcher.
+. The fetcher receives the FetchInput.
+. The fetcher emits a candidate: `Item A`.
+. The controller receives the candidate and stores it in the SDK CrawlDB. Mandatory fields are set to the item metadata, but the `blockId` field is not set.
+. Later, the candidate Item A is selected by the SDK controller, which sends it to the fetcher.
+. The fetcher receives the candidate and processes it.
+. The fetcher emits a `Document` from the candidate.
+. The fetcher emits a `FetchResult` to the SDK controller.
+. The SDK controller receives both the Document and the FetchResult.
+.. If processing the Document, the item status is updated to Document in SDK CrawlDB.
+.. If processing the FetchResult, the item's blockId is set to the current job's blockId: `N8W12TQWK3`.
```

Comment:
Is this blockId sample value used anywhere else in this doc? If not, we can just remove that.